### ISSUE TITLE
sslv3 alert handshake failure

### ISSUE THREAD

**Issue by bhagatsajan0073** [scrapy/scrapy#2324]
Error: [('SSL routines', 'SSL23_GET_SERVER_HELLO', 'sslv3 alert handshake failure')]
<img width="809" alt="sslv3" src="https://cloud.githubusercontent.com/assets/10372487/19321896/75496a66-90d4-11e6-8c1c-9018bcaaecb4.PNG">

**Comment by redapple** [comment]
hi @bhagatsajan0073 , thanks for reporting.
can you give us the URL for which you are getting this error?
Are you able to test this patch https://github.com/scrapy/scrapy/pull/2314?

**Comment by bhagatsajan0073** [comment]
yes i tested the patch but i'm getting the same error 
url : https://www.vcpi.com

**Comment by redapple** [comment]
It works for me with scrapy 1.2, without the fix with more OpenSSL ciphers:

```
(scrapy12) paul@paul-SATELLITE-R830:~/src/scrapy$ scrapy version -v
Scrapy    : 1.2.0
lxml      : 3.6.4.0
libxml2   : 2.9.4
Twisted   : 16.4.1
Python    : 2.7.12 (default, Jul  1 2016, 15:12:24) - [GCC 5.4.0 20160609]
pyOpenSSL : 16.1.0 (OpenSSL 1.0.2g  1 Mar 2016)
Platform  : Linux-4.4.0-42-generic-x86_64-with-Ubuntu-16.04-xenial

$ scrapy shell https://www.vcpi.com/
2016-10-13 12:09:44 [scrapy] INFO: Scrapy 1.2.0 started (bot: scrapybot)
(...)
2016-10-13 12:09:46 [scrapy] DEBUG: Crawled (200) <GET https://www.vcpi.com/> (referer: None)
>>> 

```

What's your output of `scrapy version -v`?
Could you be using an older OpenSSL or Twisted version?

**Comment by bhagatsajan0073** [comment]
$ scrapy version -v
Scrapy    : 1.2.0
lxml      : 3.6.4.0
libxml2   : 2.9.4
Twisted   : 16.4.1
Python    : 2.7.9 (default, Dec 10 2014, 12:28:03) [MSC v.1500 64 bit (AMD64)]
pyOpenSSL : 16.1.0 (OpenSSL 1.0.2j  26 Sep 2016)
Platform  : Windows-8-6.2.9200

**Comment by redapple** [comment]
are you able to connect using `openssl s_client -connect www.vcpi.com:443`?

**Comment by redapple** [comment]
This is what I get myself:

```
$ openssl s_client -connect www.vcpi.com:443 -servername  www.vcpi.com
CONNECTED(00000003)
depth=2 O = Digital Signature Trust Co., CN = DST Root CA X3
verify return:1
depth=1 C = US, O = Let's Encrypt, CN = Let's Encrypt Authority X3
verify return:1
depth=0 CN = secure0027.hubspot.com
verify return:1
---
Certificate chain
 0 s:/CN=secure0027.hubspot.com
   i:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
 1 s:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIM4zCCC8ugAwIBAgISAz38Fn91ecfMb/l1H33rv9ABMA0GCSqGSIb3DQEBCwUA
...
Z7Q0vqhGJQpd7i1/XwsfVy8fgLvHOn3PONV/Jo4KMdDAxoiLD9oZ
-----END CERTIFICATE-----
subject=/CN=secure0027.hubspot.com
issuer=/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 4987 bytes and written 452 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: F0E715961...
    Session-ID-ctx: 
    Master-Key: A129C0B04390963703...
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1476363122
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---
closed
```

**Comment by bhagatsajan0073** [comment]
$ openssl s_client -connect www.vcpi.com:443 -servername www.vcpi.com
CONNECTED(000001B4)
---
Certificate chain
 0 s:/CN=secure0027.hubspot.com
   i:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
 1 s:/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
   i:/O=Digital Signature Trust Co./CN=DST Root CA X3
---
Server certificate
-----BEGIN CERTIFICATE-----
MIIM4zCCC8ugAwIBAgISAz38Fn91ecfMb/l1H33rv9ABMA0GCSqGSIb3DQEBCwUA
......................
Z5s58kJo0BogR026ka5gZhvoGYxha1eTJKDfbgTLbHFMuaDFImljLKHA9JBUJgwa
Z7Q0vqhGJQpd7i1/XwsfVy8fgLvHOn3PONV/Jo4KMdDAxoiLD9oZ
-----END CERTIFICATE-----
subject=/CN=secure0027.hubspot.com
issuer=/C=US/O=Let's Encrypt/CN=Let's Encrypt Authority X3
---
No client certificate CA names sent
Peer signing digest: SHA512
Server Temp Key: ECDH, P-256, 256 bits
---
SSL handshake has read 4987 bytes and written 455 bytes
---
New, TLSv1/SSLv3, Cipher is ECDHE-RSA-AES256-GCM-SHA384
Server public key is 2048 bit
Secure Renegotiation IS supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : ECDHE-RSA-AES256-GCM-SHA384
    Session-ID: CD0EB5A5E93B309........
    Session-ID-ctx:
    Master-Key: 2C9325D28853FDFBB5.......
    Key-Arg   : None
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1476363748
    Timeout   : 300 (sec)
    Verify return code: 0 (ok)
---

**Comment by redapple** [comment]
ok, then I don't know what may be happening here for you and scrapy.
I cannot reproduce it. I don't have a Windows machine at hand.
Can you try in a virtualenv?

**Comment by redapple** [comment]
@bhagatsajan0073 , there's one 301 response in your stats. What does it correspond to? Are you being redirected to another page or website?

**Comment by bhagatsajan0073** [comment]
https://www.vcpi.com to this url

**Comment by redapple** [comment]
@bhagatsajan0073 , have you tried in a virtualenv?
Otherwise I can only suggest using something like tcpdump or wireshark to understand why the handshake fails

**Comment by bhagatsajan0073** [comment]
i'll see with wireshark

**Comment by redapple** [comment]
@bhagatsajan0073 , did you make progress in the investigation?

**Comment by jxltom** [comment]
I can same error.

```
scrapy version -v
Scrapy    : 1.3.1
lxml      : 3.7.2.0
libxml2   : 2.9.3
cssselect : 1.0.1
parsel    : 1.1.0
w3lib     : 1.17.0
Twisted   : 17.1.0
Python    : 3.4.6 (default, Feb 10 2017, 02:06:15) - [GCC 4.9.2]
pyOpenSSL : 16.2.0 (OpenSSL 1.0.1t  3 May 2016)
Platform  : Linux-4.4.0-59-generic-x86_64-with-debian-8.7
```

Error is raised when ```scrapy shell https://doc.scrapy.org/en/latest/```

@redapple Any thoughts? Thanks.

**Comment by jxltom** [comment]
By the way, everyting is fine after I uninstall twisted 17.1.0 and install Twisted 16.6.0 with pip.

```shell
# scrapy version -v
Scrapy    : 1.3.1
lxml      : 3.7.2.0
libxml2   : 2.9.3
cssselect : 1.0.1
parsel    : 1.1.0
w3lib     : 1.17.0
Twisted   : 16.6.0
Python    : 3.4.6 (default, Feb 10 2017, 02:06:15) - [GCC 4.9.2]
pyOpenSSL : 16.2.0 (OpenSSL 1.0.1t  3 May 2016)
Platform  : Linux-4.4.0-59-generic-x86_64-with-debian-8.7
```

**Comment by jxltom** [comment]
OK, I found the issue is discussed in #2555 and solved in #2558. Update scrapy to 1.3.2 solved my problem.

**Comment by cathalgarvey** [comment]
Thanks for checking back in on this @jxltom - closing this as the evolution of SSL/TLS-related issues has moved to newer issues. Thanks for your contribution to Scrapy. :)
